internal/cli/auth/login.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2022 MongoDB Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/mongodb/mongocli/internal/cli/require"
+	"github.com/mongodb/mongocli/internal/config"
+	"github.com/mongodb/mongocli/internal/oauth"
+	"github.com/pkg/browser"
+	"github.com/spf13/cobra"
+	"go.mongodb.org/atlas/auth"
+	atlas "go.mongodb.org/atlas/mongodbatlas"
+)
+
+//go:generate mockgen -destination=../../mocks/mock_login.go -package=mocks github.com/mongodb/mongocli/internal/cli/auth Authenticator
+
+type Authenticator interface {
+	RequestCode(context.Context) (*auth.DeviceCode, *atlas.Response, error)
+	PollToken(context.Context, *auth.DeviceCode) (*auth.Token, *atlas.Response, error)
+}
+
+type loginOpts struct {
+	OutWriter      io.Writer
+	AuthToken      string
+	RefreshToken   string
+	OpsManagerURL  string
+	ProjectID      string
+	OrgID          string
+	MongoShellPath string
+	isGov          bool
+	isCloudManager bool
+	noBrowser      bool
+	config         config.SetSaver
+	flow           Authenticator
+}
+
+func (opts *loginOpts) initFlow() error {
+	var err error
+	opts.flow, err = oauth.FlowWithConfig(config.Default())
+	return err
+}
+
+func (opts *loginOpts) SetUpAccess() {
+	if opts.isGov {
+		opts.config.Set("service", config.CloudGovService)
+	} else if opts.isCloudManager {
+		opts.config.Set("service", config.CloudManagerService)
+	}
+
+	if opts.AuthToken != "" {
+		opts.config.Set("auth_token", opts.AuthToken)
+	}
+	if opts.RefreshToken != "" {
+		opts.config.Set("refresh_token", opts.RefreshToken)
+	}
+	if opts.OpsManagerURL != "" {
+		opts.config.Set("ops_manager_url", opts.OpsManagerURL)
+	}
+}
+
+func (opts *loginOpts) Run(ctx context.Context) error {
+	code, _, err := opts.flow.RequestCode(ctx)
+	if err != nil {
+		return err
+	}
+
+	codeDuration := time.Duration(code.ExpiresIn) * time.Second
+	_, _ = fmt.Fprintf(opts.OutWriter, `
+Here is your one-time code: %s-%s
```

Comment:
this is the design but after using it starting with `here is...` felt quite abrupt as a user so open for suggestions @mongodb/docs-cloud-team

Reply:
We could make it a little friendlier with sequential language maybe? See comment below with suggestion.